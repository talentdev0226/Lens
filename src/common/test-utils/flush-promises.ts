/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { setImmediate } from "timers";

export const flushPromises = () => new Promise(setImmediate);
