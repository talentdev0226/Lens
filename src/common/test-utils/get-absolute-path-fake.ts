/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { GetAbsolutePath } from "../path/get-absolute-path.injectable";

export const getAbsolutePathFake: GetAbsolutePath = (...args) => {
  const maybeAbsolutePath = args.join("/");

  if (isAbsolutePath(maybeAbsolutePath)) {
    return maybeAbsolutePath;
  }

  return `/some-absolute-root-directory/${maybeAbsolutePath}`;
};

const isAbsolutePath = (path: string) => path.startsWith("/");
