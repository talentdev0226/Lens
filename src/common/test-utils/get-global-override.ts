/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { Injectable } from "@ogre-tools/injectable";

export interface GlobalOverride {
  injectable: Injectable<any, any, any>;
  overridingInstantiate: any;
}

export const getGlobalOverride = <T extends Injectable<any, any, any>>(
  injectable: T,
  overridingInstantiate: T["instantiate"],
) => ({
    injectable,
    overridingInstantiate,
  });
