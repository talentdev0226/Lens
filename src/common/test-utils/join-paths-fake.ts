/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { JoinPaths } from "../path/join-paths.injectable";

export const joinPathsFake: JoinPaths = (...args) => args.join("/");
