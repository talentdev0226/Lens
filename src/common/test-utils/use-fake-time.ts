/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { act } from "@testing-library/react";

let usingFakeTime = false;

export const advanceFakeTime = (milliseconds: number) => {
  if (!usingFakeTime) {
    throw new Error("Tried to advance fake time but it was not enabled. Call useFakeTime() first.");
  }

  act(() => {
    jest.advanceTimersByTime(milliseconds);
  });
};

export const useFakeTime = (dateTime = "2015-10-21T07:28:00Z") => {
  usingFakeTime = true;

  jest.useFakeTimers();

  jest.setSystemTime(new Date(dateTime));
};
