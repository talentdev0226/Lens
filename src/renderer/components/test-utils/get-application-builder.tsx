/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import rendererExtensionsInjectable from "../../../extensions/renderer-extensions.injectable";
import currentlyInClusterFrameInjectable from "../../routes/currently-in-cluster-frame.injectable";
import { extensionRegistratorInjectionToken } from "../../../extensions/extension-loader/extension-registrator-injection-token";
import { computed, IObservableArray, observable, runInAction } from "mobx";
import { renderFor } from "./renderFor";
import observableHistoryInjectable from "../../navigation/observable-history.injectable";
import React from "react";
import { Router } from "react-router";
import { Observer } from "mobx-react";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";
import allowedResourcesInjectable from "../../../common/cluster-store/allowed-resources.injectable";
import { fireEvent, RenderResult } from "@testing-library/react";
import type { KubeResource } from "../../../common/rbac";
import directoryForLensLocalStorageInjectable from "../../../common/directory-for-lens-local-storage/directory-for-lens-local-storage.injectable";
import { Sidebar } from "../layout/sidebar";
import { getDisForUnitTesting } from "../../../test-utils/get-dis-for-unit-testing";
import type { DiContainer } from "@ogre-tools/injectable";
import clusterStoreInjectable from "../../../common/cluster-store/cluster-store.injectable";
import type { ClusterStore } from "../../../common/cluster-store/cluster-store";
import mainExtensionsInjectable from "../../../extensions/main-extensions.injectable";
import type { LensMainExtension } from "../../../extensions/lens-main-extension";
import currentRouteComponentInjectable from "../../routes/current-route-component.injectable";
import { pipeline } from "@ogre-tools/fp";
import { flatMap, compact, join, get, filter } from "lodash/fp";
import preferenceNavigationItemsInjectable from "../+preferences/preferences-navigation/preference-navigation-items.injectable";
import navigateToPreferencesInjectable from "../../../common/front-end-routing/routes/preferences/navigate-to-preferences.injectable";
import applicationMenuItemsInjectable, { MenuItemOpts } from "../../../main/menu/application-menu-items.injectable";

type Callback = (dis: DiContainers) => void | Promise<void>;

export interface ApplicationBuilder {
  dis: DiContainers;
  setEnvironmentToClusterFrame: () => ApplicationBuilder;
  addExtensions: (...extensions: LensRendererExtension[]) => Promise<ApplicationBuilder>;
  allowKubeResource: (resourceName: KubeResource) => ApplicationBuilder;
  beforeSetups: (callback: Callback) => ApplicationBuilder;
  beforeRender: (callback: Callback) => ApplicationBuilder;
  render: () => Promise<RenderResult>;

  applicationMenu: {
    click: (path: string) => void;
  };

  preferences: {
    close: () => void;
    navigate: () => void;
    navigation: {
      click: (id: string) => void;
    };
  };
}

interface DiContainers {
  rendererDi: DiContainer;
  mainDi: DiContainer;
}

interface Environment {
  renderSidebar: () => React.ReactNode;
  onAllowKubeResource: () => void;
}

export const getApplicationBuilder = () => {
  const { rendererDi, mainDi, runSetups } = getDisForUnitTesting({
    doGeneralOverrides: true,
  });

  const dis = { rendererDi, mainDi };

  const clusterStoreStub = {
    getById: (): null => null,
  } as unknown as ClusterStore;

  rendererDi.override(clusterStoreInjectable, () => clusterStoreStub);
  mainDi.override(clusterStoreInjectable, () => clusterStoreStub);

  const beforeSetupsCallbacks: Callback[] = [];
  const beforeRenderCallbacks: Callback[] = [];

  const extensionsState = observable.array<LensRendererExtension>();

  rendererDi.override(subscribeStoresInjectable, () => () => () => {});

  const environments: Record<string, Environment> = {
    application: {
      renderSidebar: () => null,

      onAllowKubeResource: () => {
        throw new Error(
          "Tried to allow kube resource when environment is not cluster frame.",
        );
      },
    },

    clusterFrame: {
      renderSidebar: () => <Sidebar />,
      onAllowKubeResource: () => {},
    },
  };

  let environment = environments.application;

  rendererDi.override(
    currentlyInClusterFrameInjectable,
    () => environment === environments.clusterFrame,
  );

  rendererDi.override(rendererExtensionsInjectable, () =>
    computed(() => extensionsState),
  );

  mainDi.override(mainExtensionsInjectable, () =>
    computed((): LensMainExtension[] => []),
  );

  let allowedResourcesState: IObservableArray<KubeResource>;
  let rendered: RenderResult;

  const builder: ApplicationBuilder = {
    dis,

    applicationMenu: {
      click: (path: string) => {
        const applicationMenuItems = mainDi.inject(
          applicationMenuItemsInjectable,
        );

        const menuItems = pipeline(
          applicationMenuItems.get(),
          flatMap(toFlatChildren(null)),
          filter((menuItem) => !!menuItem.click),
        );

        const menuItem = menuItems.find((menuItem) => menuItem.path === path);

        if (!menuItem) {
          const availableIds = menuItems.map(get("path")).join('", "');

          throw new Error(
            `Tried to click application menu item with ID "${path}" which does not exist. Available IDs are: "${availableIds}"`,
          );
        }

        menuItem.click(undefined, undefined, undefined);
      },
    },

    preferences: {
      close: () => {
        const link = rendered.getByTestId("close-preferences");

        fireEvent.click(link);
      },

      navigate: () => {
        const navigateToPreferences = rendererDi.inject(navigateToPreferencesInjectable);

        navigateToPreferences();
      },

      navigation: {
        click: (id: string) => {
          const link = rendered.queryByTestId(`tab-link-for-${id}`);

          if (!link) {
            const preferencesNavigationItems = rendererDi.inject(
              preferenceNavigationItemsInjectable,
            );

            const availableIds = preferencesNavigationItems
              .get()
              .map(get("id"));

            throw new Error(
              `Tried to click navigation item "${id}" which does not exist in preferences. Available IDs are "${availableIds.join(
                '", "',
              )}"`,
            );
          }

          fireEvent.click(link);
        },
      },
    },

    setEnvironmentToClusterFrame: () => {
      environment = environments.clusterFrame;

      allowedResourcesState = observable.array();

      rendererDi.override(allowedResourcesInjectable, () =>
        computed(() => new Set([...allowedResourcesState])),
      );

      rendererDi.override(
        directoryForLensLocalStorageInjectable,
        () => "/irrelevant",
      );

      return builder;
    },

    addExtensions: async (...extensions) => {
      const extensionRegistrators = rendererDi.injectMany(
        extensionRegistratorInjectionToken,
      );

      const addAndEnableExtensions = async () => {
        const registratorPromises = extensions.flatMap((extension) =>
          extensionRegistrators.map((registrator) => registrator(extension, 1)),
        );

        await Promise.all(registratorPromises);

        runInAction(() => {
          extensions.forEach((extension) => {
            extensionsState.push(extension);
          });
        });
      };

      if (rendered) {
        await addAndEnableExtensions();
      } else {
        builder.beforeRender(addAndEnableExtensions);
      }

      return builder;
    },

    allowKubeResource: (resourceName) => {
      environment.onAllowKubeResource();

      runInAction(() => {
        allowedResourcesState.push(resourceName);
      });

      return builder;
    },

    beforeSetups(callback: (dis: DiContainers) => void) {
      beforeSetupsCallbacks.push(callback);

      return builder;
    },

    beforeRender(callback: (dis: DiContainers) => void) {
      beforeRenderCallbacks.push(callback);

      return builder;
    },

    async render() {
      for (const callback of beforeSetupsCallbacks) {
        await callback(dis);
      }

      await runSetups();

      const render = renderFor(rendererDi);

      const history = rendererDi.inject(observableHistoryInjectable);

      const currentRouteComponent = rendererDi.inject(
        currentRouteComponentInjectable,
      );

      for (const callback of beforeRenderCallbacks) {
        await callback(dis);
      }

      rendered = render(
        <Router history={history}>
          {environment.renderSidebar()}

          <Observer>
            {() => {
              const Component = currentRouteComponent.get();

              if (!Component) {
                return null;
              }

              return <Component />;
            }}
          </Observer>
        </Router>,
      );

      return rendered;
    },
  };

  return builder;
};

const toFlatChildren =
  (parentId: string) =>
    ({
      submenu = [],
      ...menuItem
    }: MenuItemOpts): (MenuItemOpts & { path: string })[] =>
      [
        {
          ...menuItem,
          path: pipeline([parentId, menuItem.id], compact, join(".")),
        },
        ...submenu.flatMap(toFlatChildren(menuItem.id)),
      ];
