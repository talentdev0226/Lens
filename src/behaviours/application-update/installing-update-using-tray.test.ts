/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { ApplicationBuilder } from "../../renderer/components/test-utils/get-application-builder";
import { getApplicationBuilder } from "../../renderer/components/test-utils/get-application-builder";
import type { RenderResult } from "@testing-library/react";
import electronUpdaterIsActiveInjectable from "../../main/electron-app/features/electron-updater-is-active.injectable";
import publishIsConfiguredInjectable from "../../main/application-update/publish-is-configured.injectable";
import type { CheckForPlatformUpdates } from "../../main/application-update/check-for-platform-updates/check-for-platform-updates.injectable";
import checkForPlatformUpdatesInjectable from "../../main/application-update/check-for-platform-updates/check-for-platform-updates.injectable";
import type { AsyncFnMock } from "@async-fn/jest";
import asyncFn from "@async-fn/jest";
import type { DownloadPlatformUpdate } from "../../main/application-update/download-platform-update/download-platform-update.injectable";
import downloadPlatformUpdateInjectable from "../../main/application-update/download-platform-update/download-platform-update.injectable";
import closeAllWindowsInjectable from "../../main/start-main-application/lens-window/hide-all-windows/close-all-windows.injectable";
import applicationWindowInjectable from "../../main/start-main-application/lens-window/application-window/application-window.injectable";
import type { LensWindow } from "../../main/start-main-application/lens-window/application-window/lens-window-injection-token";

describe("installing update using tray", () => {
  let applicationBuilder: ApplicationBuilder;
  let checkForPlatformUpdatesMock: AsyncFnMock<CheckForPlatformUpdates>;
  let downloadPlatformUpdateMock: AsyncFnMock<DownloadPlatformUpdate>;

  beforeEach(() => {
    applicationBuilder = getApplicationBuilder();

    applicationBuilder.beforeApplicationStart(({ mainDi }) => {
      checkForPlatformUpdatesMock = asyncFn();
      downloadPlatformUpdateMock = asyncFn();

      mainDi.override(
        checkForPlatformUpdatesInjectable,
        () => checkForPlatformUpdatesMock,
      );

      mainDi.override(
        downloadPlatformUpdateInjectable,
        () => downloadPlatformUpdateMock,
      );

      mainDi.override(electronUpdaterIsActiveInjectable, () => true);
      mainDi.override(publishIsConfiguredInjectable, () => true);
    });
  });

  describe("when started", () => {
    let rendered: RenderResult;

    beforeEach(async () => {
      rendered = await applicationBuilder.render();
    });

    it("renders", () => {
      expect(rendered.baseElement).toMatchSnapshot();
    });

    it("user cannot install update yet", () => {
      expect(applicationBuilder.tray.get("install-update")).toBeNull();
    });

    describe("given all application windows are closed, when checking for updates", () => {
      let applicationWindow: LensWindow;
      let closeAllWindows: () => void;

      beforeEach(() => {
        const mainDi = applicationBuilder.dis.mainDi;

        closeAllWindows = mainDi.inject(closeAllWindowsInjectable);

        applicationWindow = mainDi.inject(applicationWindowInjectable);

        closeAllWindows();

        applicationBuilder.tray.click("check-for-updates");
      });

      describe("when check for update resolves with new update", () => {
        beforeEach(async () => {
          await checkForPlatformUpdatesMock.resolve({
            updateWasDiscovered: true,
            version: "some-version",
          });
        });

        it("does not show application window yet", () => {
          expect(applicationWindow.isVisible).toBe(false);
        });

        describe("when download of update resolves with success", () => {
          beforeEach(async () => {

            await downloadPlatformUpdateMock.resolve({ downloadWasSuccessful: true });
          });

          it("shows the application window", () => {
            expect(applicationWindow.isVisible).toBe(true);
          });

          it("given closing application window again and checking for updates again using tray, when check resolves with same version that was earlier downloaded, shows the application window", async () => {
            closeAllWindows();

            applicationBuilder.tray.click("check-for-updates");

            await checkForPlatformUpdatesMock.resolve({
              updateWasDiscovered: true,
              version: "some-version",
            });

            expect(applicationWindow.isVisible).toBe(true);
          });
        });

        it("when download of update resolves with failure, does not show the application window", async () => {
          await downloadPlatformUpdateMock.resolve({ downloadWasSuccessful: false });

          expect(applicationWindow.isVisible).toBe(false);
        });
      });

      it("when process resolves without new update, does not show the application window", async () => {
        await checkForPlatformUpdatesMock.resolve({
          updateWasDiscovered: false,
        });

        expect(applicationWindow.isVisible).toBe(false);
      });
    });

    describe("when user checks for updates using tray", () => {
      beforeEach(() => {
        applicationBuilder.tray.click("check-for-updates");
      });

      it("user cannot check for updates again", () => {
        expect(
          applicationBuilder.tray.get("check-for-updates")?.enabled,
        ).toBe(false);
      });

      it("name of tray item for checking updates indicates that checking is happening", () => {
        expect(
          applicationBuilder.tray.get("check-for-updates")?.label,
        ).toBe("Checking for updates...");
      });

      it("user cannot install update yet", () => {
        expect(applicationBuilder.tray.get("install-update")).toBeNull();
      });

      it("renders", () => {
        expect(rendered.baseElement).toMatchSnapshot();
      });

      describe("when no new update is discovered", () => {
        beforeEach(async () => {
          await checkForPlatformUpdatesMock.resolve({
            updateWasDiscovered: false,
          });
        });

        it("user cannot install update", () => {
          expect(applicationBuilder.tray.get("install-update")).toBeNull();
        });

        it("user can check for updates again", () => {
          expect(
            applicationBuilder.tray.get("check-for-updates")?.enabled,
          ).toBe(true);
        });

        it("name of tray item for checking updates no longer indicates that checking is happening", () => {
          expect(
            applicationBuilder.tray.get("check-for-updates")?.label,
          ).toBe("Check for updates");
        });

        it("renders", () => {
          expect(rendered.baseElement).toMatchSnapshot();
        });
      });

      describe("when new update is discovered", () => {
        beforeEach(async () => {
          await checkForPlatformUpdatesMock.resolve({
            updateWasDiscovered: true,
            version: "some-version",
          });
        });

        it("user cannot check for updates again yet", () => {
          expect(
            applicationBuilder.tray.get("check-for-updates")?.enabled,
          ).toBe(false);
        });

        it("name of tray item for checking updates indicates that downloading is happening", () => {
          expect(
            applicationBuilder.tray.get("check-for-updates")?.label,
          ).toBe("Downloading update some-version (0%)...");
        });

        it("when download progresses with decimals, percentage increases as integers", () => {
          downloadPlatformUpdateMock.mock.calls[0][0]({ percentage: 42.424242 });

          expect(
            applicationBuilder.tray.get("check-for-updates")?.label,
          ).toBe("Downloading update some-version (42%)...");
        });

        it("user still cannot install update", () => {
          expect(applicationBuilder.tray.get("install-update")).toBeNull();
        });

        it("renders", () => {
          expect(rendered.baseElement).toMatchSnapshot();
        });

        describe("when download fails", () => {
          beforeEach(async () => {
            await downloadPlatformUpdateMock.resolve({ downloadWasSuccessful: false });
          });

          it("user cannot install update", () => {
            expect(
              applicationBuilder.tray.get("install-update"),
            ).toBeNull();
          });

          it("user can check for updates again", () => {
            expect(
              applicationBuilder.tray.get("check-for-updates")?.enabled,
            ).toBe(true);
          });

          it("name of tray item for checking updates no longer indicates that downloading is happening", () => {
            expect(
              applicationBuilder.tray.get("check-for-updates")?.label,
            ).toBe("Check for updates");
          });

          it("renders", () => {
            expect(rendered.baseElement).toMatchSnapshot();
          });
        });

        describe("when download succeeds", () => {
          beforeEach(async () => {
            await downloadPlatformUpdateMock.resolve({ downloadWasSuccessful: true });
          });

          it("user can install update", () => {
            expect(
              applicationBuilder.tray.get("install-update")?.label,
            ).toBe("Install update some-version");
          });

          it("user can check for updates again", () => {
            expect(
              applicationBuilder.tray.get("check-for-updates")?.enabled,
            ).toBe(true);
          });

          it("name of tray item for checking updates no longer indicates that downloading is happening", () => {
            expect(
              applicationBuilder.tray.get("check-for-updates")?.label,
            ).toBe("Check for updates");
          });

          it("renders", () => {
            expect(rendered.baseElement).toMatchSnapshot();
          });
        });
      });
    });
  });
});
