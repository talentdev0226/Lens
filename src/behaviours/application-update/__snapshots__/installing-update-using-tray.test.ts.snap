// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`installing update using tray when started renders 1`] = `
<body>
  <div>
    <div
      class="topBar"
    >
      <div
        class="items"
      >
        <i
          class="Icon material interactive focusable"
          data-testid="home-button"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="home"
          >
            home
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-back"
        >
          <span
            class="icon"
            data-icon-name="arrow_back"
          >
            arrow_back
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-forward"
        >
          <span
            class="icon"
            data-icon-name="arrow_forward"
          >
            arrow_forward
          </span>
        </i>
      </div>
      <div
        class="items"
      />
    </div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
    <div
      class="Notifications flex column align-flex-end"
    />
  </div>
</body>
`;

exports[`installing update using tray when started when user checks for updates using tray renders 1`] = `
<body>
  <div>
    <div
      class="topBar"
    >
      <div
        class="items"
      >
        <i
          class="Icon material interactive focusable"
          data-testid="home-button"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="home"
          >
            home
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-back"
        >
          <span
            class="icon"
            data-icon-name="arrow_back"
          >
            arrow_back
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-forward"
        >
          <span
            class="icon"
            data-icon-name="arrow_forward"
          >
            arrow_forward
          </span>
        </i>
      </div>
      <div
        class="items"
      />
    </div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
    <div
      class="Notifications flex column align-flex-end"
    >
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          Checking for updates...
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-notification_43"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`installing update using tray when started when user checks for updates using tray when new update is discovered renders 1`] = `
<body>
  <div>
    <div
      class="topBar"
    >
      <div
        class="items"
      >
        <i
          class="Icon material interactive focusable"
          data-testid="home-button"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="home"
          >
            home
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-back"
        >
          <span
            class="icon"
            data-icon-name="arrow_back"
          >
            arrow_back
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-forward"
        >
          <span
            class="icon"
            data-icon-name="arrow_forward"
          >
            arrow_forward
          </span>
        </i>
      </div>
      <div
        class="items"
      />
    </div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
    <div
      class="Notifications flex column align-flex-end"
    >
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          Checking for updates...
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-notification_181"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          Download for version some-version started...
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-notification_184"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`installing update using tray when started when user checks for updates using tray when new update is discovered when download fails renders 1`] = `
<body>
  <div>
    <div
      class="topBar"
    >
      <div
        class="items"
      >
        <i
          class="Icon material interactive focusable"
          data-testid="home-button"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="home"
          >
            home
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-back"
        >
          <span
            class="icon"
            data-icon-name="arrow_back"
          >
            arrow_back
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-forward"
        >
          <span
            class="icon"
            data-icon-name="arrow_forward"
          >
            arrow_forward
          </span>
        </i>
        <button
          class="updateButton"
          data-testid="update-button"
          data-warning-level="light"
          id="update-lens-button"
        >
          Update
          <i
            class="Icon icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </button>
      </div>
      <div
        class="items"
      />
    </div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
    <div
      class="Notifications flex column align-flex-end"
    >
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          Checking for updates...
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-notification_268"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          Download for version some-version started...
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-notification_271"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          Download of update failed
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-notification_277"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`installing update using tray when started when user checks for updates using tray when new update is discovered when download succeeds renders 1`] = `
<body>
  <div>
    <div
      class="topBar"
    >
      <div
        class="items"
      >
        <i
          class="Icon material interactive focusable"
          data-testid="home-button"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="home"
          >
            home
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-back"
        >
          <span
            class="icon"
            data-icon-name="arrow_back"
          >
            arrow_back
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-forward"
        >
          <span
            class="icon"
            data-icon-name="arrow_forward"
          >
            arrow_forward
          </span>
        </i>
        <button
          class="updateButton"
          data-testid="update-button"
          data-warning-level="light"
          id="update-lens-button"
        >
          Update
          <i
            class="Icon icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="arrow_drop_down"
            >
              arrow_drop_down
            </span>
          </i>
        </button>
      </div>
      <div
        class="items"
      />
    </div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
    <div
      class="Notifications flex column align-flex-end"
    >
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          Checking for updates...
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-notification_335"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          Download for version some-version started...
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-notification_338"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`installing update using tray when started when user checks for updates using tray when no new update is discovered renders 1`] = `
<body>
  <div>
    <div
      class="topBar"
    >
      <div
        class="items"
      >
        <i
          class="Icon material interactive focusable"
          data-testid="home-button"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="home"
          >
            home
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-back"
        >
          <span
            class="icon"
            data-icon-name="arrow_back"
          >
            arrow_back
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-forward"
        >
          <span
            class="icon"
            data-icon-name="arrow_forward"
          >
            arrow_forward
          </span>
        </i>
      </div>
      <div
        class="items"
      />
    </div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
    <div
      class="Notifications flex column align-flex-end"
    >
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          Checking for updates...
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-notification_104"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          No new updates available
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-notification_107"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
    </div>
  </div>
</body>
`;
